java/sage/epg/sd/SDRecommended.java

```diff
@@ -0,0 +1,470 @@
+        if (program != null && !program.hasImageArtwork())
+        {
+          editorials.remove(i--);
+          continue;
+        }
+      }
+
+      query.add(programId);
+      if (query.size() >= limit)
+      {
+        while (i < editorials.size())
+        {
+          editorials.remove(i);
+        }
+        break;
+      }
+    }
+
+    SDProgram programs[] = SDRipper.ensureSession().getPrograms(query);
+    for (int i = 0; i < programs.length; i++)
+    {
+      SDEditorial editorial = editorials.get(i);
+      editorial.setProgram(programs[i]);
+    }
+
+    if (Sage.DBG) System.out.println("SDEPG Recommended: " + editorials.toString());
+  }
+
+  public static List<SDEditorial> getUsable(List<SDEditorial> editorials)
+  {
+    int alreadyExists = 0;
+    List<SDEditorial> returnValues = new ArrayList<>();
+    Agent favorites[] = Wizard.getInstance().getFavorites();
+    TVEditorial currentEditorials[] = Wizard.getInstance().getEditorials();
+    for (SDEditorial editorial : editorials)
+    {
+      SDRecommendation recommendation = editorial.getRecommendation();
+      String title = recommendation.getTitle120();
+      if (title == null || title.length() == 0)
+        continue;
+      String programId = recommendation.getProgramId();
+      if (programId == null || programId.length() == 0)
+        continue;
+
+      // Don't recommend movies that we have already seen or don't exist in the Wizard.
+      if (programId.startsWith("MV"))
+      {
+        Show show;
+        if ((show = Wizard.getInstance().getShowForExternalID(programId)) == null || show.isWatched())
+          continue;
+        String newId = SDUtils.fromProgramToSageTV(programId);
+        if (newId.length() == 12 &&
+          (show = Wizard.getInstance().getShowForExternalID(programId)) == null || show.isWatched())
+          continue;
+      }
+
+      // Don't recommend shows that are clearly already favorites. We might get a show that changed
+      // titles or would have been recorded due to other rules, but we aren't too concerned about
+      // that.
+      boolean skip = false;
+      for (Agent favorite : favorites)
+      {
+        if (favorite.getTitle().equals(title))
+        {
+          skip = true;
+          break;
+        }
+      }
+      if (skip)
+        continue;
+      // Sometimes we will find a different external ID for a show that's already in the editorials.
+      // As that editorial expires, if the show is still a suggestion, it will be re-added with a
+      // different external ID.
+      for (TVEditorial currentEditorial : currentEditorials)
+      {
+        if (currentEditorial.getTitle().equals(title))
+        {
+          alreadyExists++;
+          skip = true;
+          break;
+        }
+      }
+      if (skip)
+        continue;
+
+      // Check if we could't record this show if we wanted to.
+      Show shows[] = Wizard.getInstance().getShowsForExternalIDPrefix(programId);
```

Comment:
The point I was making is that when you pass a full programID into getShowsForExternalIDPrefix...it's the same as if you called getShowForExternalID because that method doesn't turn it into a prefix. Below in your code, you turn them into prefixes and call the prefix method...so you should be applying the same logic up here (if you add some debug logging, I'd expect you to only get one result returned from these first prefix calls)